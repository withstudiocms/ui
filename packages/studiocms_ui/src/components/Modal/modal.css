.sui-modal {
	border: 1px solid hsl(var(--border));
	border-radius: var(--radius-md);
	padding: 1.5rem;
	box-shadow: 0px 6px 8px hsl(var(--shadow));
	animation: hide .25s ease;
	overflow: visible;
	margin: auto;
	z-index: 50;
	max-width: calc(100% - 4rem);
}

.sui-modal.sm {
	width: 384px;
}

.sui-modal.md {
	width: 448px;
}

.sui-modal.lg {
	width: 608px;
}

.sui-modal[open] {
	animation: show .25s ease-in-out;
}

html:has(.sui-modal[open]),
body:has(.sui-modal[open]) {
	overflow: hidden;
}

.sui-modal[open]::backdrop {
	background-color: rgba(0, 0, 0, 0.75);
	animation: backdrop .3s ease-in-out forwards;
}

.sui-modal-header:has(*) {
	margin-bottom: 1rem;

	display: flex;
	flex-direction: row;
	justify-content: space-between;
	gap: 1rem;

	* {
		margin: 0;
	}
}

.x-mark-container {
	cursor: pointer;
	height: 1.5rem;
	width: 1.5rem;
	display: flex;
	align-items: center;
	justify-content: center;
	transition: background-color .15s ease;
	border-radius: var(--radius-sm);
}

.x-mark-container:hover {
	background-color: hsl(var(--default-base));
}

.x-mark-container:focus-visible {
	outline: 2px solid hsl(var(--text-normal));
	outline-offset: 2px;
}

.sui-modal-footer {
	display: none;
}

.sui-modal-footer:has(*) {
	display: flex;
	flex-direction: row;
	gap: 1rem;
	margin-top: 1rem;
	justify-content: end;
}

@keyframes hide {
	0% {
		scale: 1;
		opacity: 1;
		display: block;
	}
	100% {
		scale: 0.85;
		opacity: 0;
		display: none;
	}
}

@keyframes show {
	0% {
		scale: 0.85;
		opacity: 0;
		display: none;
	}
	100% {
		scale: 1;
		opacity: 1;
		display: block;
	}
}

@keyframes backdrop {
	0% {
		opacity: 0;
	}
	100% {
		opacity: 1;
	}
}
