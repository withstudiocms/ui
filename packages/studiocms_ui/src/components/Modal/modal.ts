class ModalHelper {
	private element: HTMLDialogElement;
	private cancelButton: HTMLButtonElement | undefined;
	private confirmButton: HTMLButtonElement | undefined;

	private isForm = false;
	private modalForm: HTMLFormElement;

	/**
	 * A helper to manage modals.
	 * @param id The ID of the modal.
	 * @param triggerID The ID of the element that should trigger the modal.
	 */
	constructor(id: string, triggerID?: string) {
		const element = document.getElementById(id) as HTMLDialogElement;

		if (!element) {
			throw new Error(`No modal with ID ${id} found.`);
		}

		this.element = element as HTMLDialogElement;
		this.modalForm = document.getElementById(`${id}-form-element`) as HTMLFormElement;

		const isDismissable = this.element.dataset.dismissable === 'true';
		const isForm = this.element.dataset.form === 'true';

		if (isDismissable) {
			this.addDismissiveClickListener();
		}

		if (isForm) this.isForm = true;

		this.addButtonListeners(id, isDismissable);

		if (triggerID) {
			this.bindTrigger(triggerID);
		}
	}

	/**
	 * A helper function which adds event listeners to the modal buttons to close the modal when clicked.
	 * @param id The ID of the modal.
	 * @param dismissable Whether the modal is dismissable.
	 */
	private addButtonListeners = (id: string, dismissable: boolean) => {
		if (
			dismissable ||
			(!this.element.dataset.hasCancelButton && !this.element.dataset.hasActionButton)
		) {
			const xMarkButton = document.getElementById(`${id}-btn-x`) as HTMLButtonElement;
			xMarkButton.addEventListener('click', this.hide);
		}

		if (!!this.element.dataset.hasCancelButton && !this.element.dataset.hasActionButton) return;

		if (this.element.dataset.hasCancelButton) {
			this.cancelButton = document.getElementById(`${id}-btn-cancel`) as HTMLButtonElement;
			this.cancelButton.addEventListener('click', this.hide);
		}

		if (this.element.dataset.hasActionButton) {
			this.confirmButton = document.getElementById(`${id}-btn-confirm`) as HTMLButtonElement;
			this.confirmButton.addEventListener('click', this.hide);
		}
	};

	/**
	 * A helper function to close the modal when the user clicks outside of it.
	 */
	private addDismissiveClickListener = () => {
		this.element.addEventListener('click', (e: MouseEvent) => {
			if (!e.target) return;

			const { left, right, top, bottom } = this.element.getBoundingClientRect();

			const clickWithinModalBox =
				e.clientX < right && e.clientX > left && e.clientY < bottom && e.clientY > top;

			if (!clickWithinModalBox) {
				this.element.close();
			}
		});
	};

	/**
	 * A function to show the modal.
	 */
	public show = () => {
		this.element.showModal();
		this.element.focus();
	};

	/**
	 * A function to hide the modal.
	 */
	public hide = () => {
		this.element.close();
	};

	/**
	 * A function to add another trigger to show the modal with.
	 * @param elementID The ID of the element that should trigger the modal when clicked.
	 */
	public bindTrigger = (elementID: string) => {
		const element = document.getElementById(elementID);

		if (!element) {
			throw new Error(`No element with ID ${elementID} found.`);
		}

		element.addEventListener('click', this.show);
	};

	/**
	 * Registers a callback for the cancel button.
	 * @param func The callback function.
	 */
	public registerCancelCallback = (func: () => void) => {
		if (!this.cancelButton) {
			throw new Error('Unable to register cancel callback without a cancel button.');
		}

		this.cancelButton.removeEventListener('click', this.hide);

		this.cancelButton.addEventListener('click', () => {
			func();
			this.hide();
		});
	};

	/**
	 * Registers a callback for the confirm button.
	 * @param func The callback function. If the modal is a form, the function will be called with 
	 * the form data as the first argument.
	 */
	public registerConfirmCallback = (func: (data?: FormData | undefined) => void) => {
		if (!this.confirmButton) {
			throw new Error('Unable to register cancel callback without a confirmation button.');
		}

		this.confirmButton.removeEventListener('click', this.hide);

		if (this.isForm) {
			this.modalForm.addEventListener('submit', (e) => {
				e.preventDefault();

				const formData = new FormData(this.modalForm);

				func(formData);
				this.hide();

				setTimeout(() => this.modalForm.reset(), 450);
			});
		} else {
			this.confirmButton.addEventListener('click', () => {
				func();
				this.hide();
			});
		}
	};
}

export { ModalHelper };
