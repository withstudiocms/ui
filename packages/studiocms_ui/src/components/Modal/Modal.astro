---
import type { StudioCMSColorway } from '../../utils/colors.js';
import Button from '../Button/Button.astro';
import Icon from '../Icon/Icon.astro';
import './modal.css';

interface ButtonType {
	label: string;
	color: StudioCMSColorway;
}

/**
 * The props for the Modal component.
 */
interface Props {
	/**
	 * The ID of the modal. Required due to the helper.
	 */
	id: string;
	/**
	 * The size of the modal. Defaults to `md`.
	 */
	size?: 'sm' | 'md' | 'lg';
	/**
	 * Whether the modal is dismissable. Defaults to `true`.
	 */
	dismissable?: boolean;
	/**
	 * The cancel button of the modal. If a string is given, the button will have the danger color and flat variant,
	 * and the string will be the label. If ButtonType is given, the button will have the color and variant specified.
	 */
	cancelButton?: string | ButtonType;
	/**
	 * The action button of the modal. If a string is given, the button will have the primary color and solid variant,
	 * and the string will be the label. If ButtonType is given, the button will have the color and variant specified.
	 */
	actionButton?: string | ButtonType;
	/**
	 * Whether the modal is a form. Defaults to `false`. When set to true, the modal helpers submit callback will include
	 * the form data.
	 */
	isForm?: boolean;
}

const {
	id,
	size = 'md',
	dismissable = true,
	isForm = false,
	cancelButton,
	actionButton,
} = Astro.props;
---
<dialog
  popover
  id={id}
  data-dismissable={`${dismissable}`}
  data-has-action-button={actionButton}
  data-has-cancel-button={cancelButton}
  class="sui-modal"
  class:list={[size]}
  data-form={isForm}
>
  <div class="sui-modal-header">
    <slot name="header" />
    {(dismissable || (!cancelButton && !actionButton)) && (
      <button class="x-mark-container" id={`${id}-btn-x`}>
        <Icon name="x-mark" width={24} height={24} class={'dismiss-icon'} />
      </button>
    )}
  </div>
  <form id={`${id}-form-element`}>
    <div class="sui-modal-body">
      <slot />
    </div>
    <div class="sui-modal-footer">
      {cancelButton && (
        <Button
          id={`${id}-btn-cancel`}
          color={typeof cancelButton === 'string' ? 'danger' : cancelButton.color} 
          variant='flat'
          type={isForm ? 'reset' : 'button'}
        >
          {typeof cancelButton === 'string' ? cancelButton : cancelButton.label}
        </Button>
      )}
      {actionButton && (
        <Button 
          id={`${id}-btn-confirm`} 
          type={'submit'} 
          color={typeof actionButton === 'string' ? 'primary' : actionButton.color} 
          variant='solid' 
          type={isForm ? 'submit' : 'button'}
        >
          {typeof actionButton === 'string' ? actionButton : actionButton.label}
        </Button>
      )}
    </div>
  </form>
</dialog>
