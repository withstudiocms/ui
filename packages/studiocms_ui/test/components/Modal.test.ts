import { describe, expect } from 'vitest';
import Modal from '../../src/components/Modal/Modal.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const loremIpsum = `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
`;

const mockProps = {
	id: 'modal-test',
};

describe('Modal Component', () => {
	test('renders Modal component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Modal, 'Modal', {
			props: mockProps,
			slots: { default: loremIpsum },
		});
		expect(result).toMatchSnapshot();
	});

	test.for([{ size: 'sm' }, { size: 'md' }, { size: 'lg' }])(
		'renders Modal with size "%s"',
		async (props, { renderComponent }) => {
			const result = await renderComponent(Modal, 'Modal', {
				props: { ...mockProps, ...props },
				slots: { default: loremIpsum },
			});
			expect(result).toMatchSnapshot();
		}
	);
});
