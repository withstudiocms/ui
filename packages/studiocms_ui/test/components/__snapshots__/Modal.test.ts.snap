// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Component > renders Modal component correctly 1`] = `
"<dialog popover id="modal-test" data-dismissable="true" class="sui-modal md" data-form="false"> <div class="sui-modal-header">  <button class="x-mark-container" id="modal-test-btn-x"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="dismiss-icon" height="24" width="24" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6 18L18 6M6 6l12 12"/></svg> </button> </div> <form id="modal-test-form-element"> <div class="sui-modal-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
 </div> <div class="sui-modal-footer">   </div> </form> </dialog>"
`;

exports[`Modal Component > renders Modal with size "{ size: 'lg' }" 1`] = `
"<dialog popover id="modal-test" data-dismissable="true" class="sui-modal lg" data-form="false"> <div class="sui-modal-header">  <button class="x-mark-container" id="modal-test-btn-x"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="dismiss-icon" height="24" width="24" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6 18L18 6M6 6l12 12"/></svg> </button> </div> <form id="modal-test-form-element"> <div class="sui-modal-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
 </div> <div class="sui-modal-footer">   </div> </form> </dialog>"
`;

exports[`Modal Component > renders Modal with size "{ size: 'md' }" 1`] = `
"<dialog popover id="modal-test" data-dismissable="true" class="sui-modal md" data-form="false"> <div class="sui-modal-header">  <button class="x-mark-container" id="modal-test-btn-x"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="dismiss-icon" height="24" width="24" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6 18L18 6M6 6l12 12"/></svg> </button> </div> <form id="modal-test-form-element"> <div class="sui-modal-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
 </div> <div class="sui-modal-footer">   </div> </form> </dialog>"
`;

exports[`Modal Component > renders Modal with size "{ size: 'sm' }" 1`] = `
"<dialog popover id="modal-test" data-dismissable="true" class="sui-modal sm" data-form="false"> <div class="sui-modal-header">  <button class="x-mark-container" id="modal-test-btn-x"> <svg style="min-width: 24px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="dismiss-icon" height="24" width="24" viewbox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6 18L18 6M6 6l12 12"/></svg> </button> </div> <form id="modal-test-form-element"> <div class="sui-modal-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
 </div> <div class="sui-modal-footer">   </div> </form> </dialog>"
`;
